### PR TITLE
title formatting is now a renderer decision

### PR DESCRIPTION
**Pull request by sabine**
This patch puts the responsibility of rendering the title (part of the "header" section) on the renderer, as suggested in the comment at https://github.com/ocaml/odoc/blob/a52dd8e960f99b38162d40311a9c4512d5741d43/src/document/generator.ml#L24

This refactoring is a prerequisite for adding more varied renderers, e.g., as requested in https://github.com/ocaml/odoc/issues/374 and as we need for the ocaml.org website's package docs browser.

`Odoc_document.Types.Page.t` now contains `preamble` and `kind` instead of `header` and the decision on whether (and how) to render a title is now made in the respective `Generator.Page.page` function of the corresponding renderer.

As I didn't change anything about the title rendering itself, I put the title rendering code shared between the different renderers in a new module `Doctree.PageTitle`. I am very open to suggestions where else to put it. Also, if there's anything I need to change about this PR, let me know.

### CODE REVIEW

**Review by Julow — APPROVED** [review]
LGTM :)

I don't know if `Doctree` is the best place but we don't have a better place at the moment. We can ask this question at if we have more shared rendering functions in the future.

**Comment by Julow** [comment]
Needs some ocamlformatting.

File: src/document/doctree.ml [hunk 1]
```diff
@@ -283,6 +283,35 @@ end = struct
...
+    let prefix s =
+      mk (Types.inline (Text (s ^ " ")) :: Codefmt.code (Codefmt.txt name)) in
+    match kind with
+    | `BaseModule -> prefix "Module"
+    | `Module -> prefix "Module"
+    | `Argument -> prefix "Parameter"
+    | `ModuleType -> prefix "Module type"
+    | `ClassType -> prefix "Class type"
+    | `Class -> prefix "Class"
+    | `Page -> []
```
Comment (dbuenzli) [inline src/document/doctree.ml:303]: Aren't you dropping the page name here ?

Comment (Julow) [reply src/document/doctree.ml:303]: The page title is embedded in the page it seems.

Comment (sabine) [reply src/document/doctree.ml:303]: I had the tests fail with a duplicate title when I kept the code as it was. The reason is that it was dead code as the title was never rendered into the header anyways, in the case of `Page` (see line 1697 here): 
https://github.com/ocaml/odoc/blob/a52dd8e960f99b38162d40311a9c4512d5741d43/src/document/generator.ml#L1691-L1698

**Comment by sabine** [comment]
Force pushed after `dune fmt`. No other changes.

**Comment by jonludlam** [comment]
Sorry it's taken a while to get to this. I'm unsure about the distinction between `` `BaseModule `` and `` `Module `` - it looks like it's to prevent putting the parent pages's title into a 'base module's title. What do you think about handling this in `make_name_from_path`? e.g.:

```ocaml
let make_name_from_path { Url.Path.name; parent; _ } =
  match parent with
  | None | Some { kind = `Page; _ } -> name
  | Some p -> Printf.sprintf "%s.%s" p.name name
```

I'm not strongly opinionated on this, so happy either way.

**Comment by sabine** [comment]
> What do you think about handling this in make_name_from_path?

Thanks for pointing that out. I took a closer look at why I added `` `BaseModule `` and what is going on in the surroundings of that.

As it stands, `Url.Path.name` is not the name, but a path fragment (it's differs from the name in that every functor parameter has their position prepended to its name), so I renamed that to `Url.Path.path_fragment`.

It turns out that `Page.title` is only used in one single place (HTML document title, all other titles were already being generated based on `Page.url`), so, for the sake of having a single source of truth, and less potential for confusion, this PR now also proposes to remove `title` from `Page`.

(I am not entirely happy with this part of the code where I have to strip the parameter number from `Url.Path.path_fragment` to get the old HTML page title: https://github.com/ocaml/odoc/pull/896/files#diff-6c455c2657c68cf89b442c7b3c19fef8503cf643a31ad0cb81d926d34e47ac10R407-R418.
An alternative to stripping the prefix from the path fragment is to keep `Url.path.name` and add the prefix for `` `Parameter `` to render the Url.Path.t is rendered to a path fragment. That would require adding an integer to `Url.path.kind`'s `` `Parameter ``.)

Opinions?

**Comment by tmattio** [comment]
Gentle ping @Julow @jonludlam, is there anything blocking the merge of the PR? This would be quite useful to make progress on the updated package documentation on OCaml.org

**Comment by jonludlam** [comment]
Looks great now, thanks! Quite happy with the addition of the integer into the `` `Parameter `` constructor.

**Pull request merged**